clash-lib/src/Clash/Core/Evaluator/Semantics.hs

```diff
@@ -0,0 +1,581 @@
+    Just b
+      | bindingSpec b /= NoInline -> do
+          v <- forceGlobal (bindingTerm b)
+
+          updateGlobal i v
+          pure (Just v)
+
+    _ -> pure Nothing
+   where
+    forceGlobal = either (withoutGlobal i . eval) pure
+
+-- | Default implementation for evaluating a literal.
+-- This simply wraps the literal up into a Value.
+--
+evaluateLiteral :: Literal -> Eval Value
+evaluateLiteral = pure . VLit
+
+-- | Default implementation for evaluating data constructors. If the
+-- constructor is not nullary, then it is eta-expanded and its eta-expanded
+-- form is evaluated to obtain a result.
+--
+evaluateDataWith
+  :: (Term -> Eval Value)
+  -> DataCon
+  -> Eval Value
+evaluateDataWith eval dc
+  | isFullyApplied (dcType dc) [] = VData dc [] <$> getLocalEnv
+  | otherwise = etaExpand (Data dc) >>= eval
+
+-- | Default implementation for evaluating primitive operations. If the primop
+-- is not nullary, then it is eta-expanded and its eta-expanded form is
+-- evaluated with the primitive evaluator to obtain a result.
+--
+evaluatePrimWith
+  :: (Term -> Eval Value)
+  -> (PrimInfo -> [Either Term Type] -> Eval Value)
+  -> PrimInfo
+  -> Eval Value
+evaluatePrimWith eval evalPrim p
+  | isFullyApplied (primType p) [] = evalPrim p []
+  | otherwise = etaExpand (Prim p) >>= eval
+
+-- | Default implementation for evaluating lambdas. As a term with a lambda
+-- at the head is already in WHNF, this simply returns the term under the
+-- lambda with the current local environment.
+--
+evaluateLam :: Id -> Term -> Eval Value
+evaluateLam i x = VLam i x <$> getLocalEnv
+
+-- | Default implementation for evaluating type lambdas. As a term with a type
+-- lambda at the head is already in WHNF, this simply returns the term under
+-- the type lambda with the current local environment.
+--
+evaluateTyLam :: TyVar -> Term -> Eval Value
+evaluateTyLam i x = VTyLam i x <$> getLocalEnv
+
+-- | Default implementation for evalating a term / type application. This
+-- checks if the arguments are applied to a data constructor or primitive, and
+-- creates a value if it is fully applied. If the arguments are applied to a
+-- function, the first argument is applied.
+--
+evaluateAppWith
+  :: (Term -> Eval Value)
+  -> (PrimInfo -> [Either Term Type] -> Eval Value)
+  -> (Value -> Either TermOrValue Type -> Eval Value)
+  -> Term
+  -> Either Term Type
+  -> Eval Value
+evaluateAppWith eval evalPrim apply x y
+  | Data dc <- f
+  = if isFullyApplied (dcType dc) args
+      then VData dc args <$> getLocalEnv
+      else etaExpand term >>= eval
+
+  | Prim p <- f
+  , nArgs  <- length . fst $ splitFunForallTy (primType p)
+  = case compare (length args) nArgs of
+      LT -> etaExpand term >>= eval
+      EQ -> evalPrim p args
+      GT -> do
+        let (pArgs, rArgs) = splitAt nArgs args
+        primRes <- evalPrim p pArgs
+        foldM apply primRes (first Left <$> rArgs)
+
+  | otherwise
+  = withIgnored $ do
```

Comment:
Add a comment why `withIgnored` is only needed here, and not for `Data` or `Prim`